diesel_codegen/src/type_inference.rs

```diff
@@ -0,0 +1,254 @@
+use syn;
+use quote;
+
+use diesel_codegen_shared::*;
+
+use util::{get_options_from_input, get_option, get_optional_option};
+
+use std::collections::HashSet;
+
+pub fn derive_infer_enums(input: syn::MacroInput) -> quote::Tokens {
+    fn bug() -> ! {
+        panic!("This is a bug. Please open a Github issue with your invocation of `infer_enums!`");
+    }
+
+    let options = get_options_from_input("infer_enum_options", &input.attrs, bug)
+        .unwrap_or_else(|| bug());
+    let database_url = get_option(&options, "database_url", bug);
+    let schema_name = get_optional_option(&options, "schema_name");
+    let types = get_optional_option(&options, "types");
+
+    infer_enums_for_schema_name(&database_url,
+                                schema_name.as_ref().map(|s| &**s),
+                                types.as_ref().map(|s| &**s),
+                                true, true)
+}
+
+fn read_type_names(csl: &str) -> HashSet<String> {
+    csl.split(",")
+        .map(|t| t.trim())
+        .map(canonicalize_pg_type_name)
+        .filter(|t| !t.is_empty()).collect()
+}
+
+fn infer_enums_for_schema_name(database_url: &str, schema_name: Option<&str>,
+                               type_list: Option<&str>, camel_case_types: bool,
+                               camel_case_variants: bool) -> quote::Tokens {
```

Comment:
Formatting is mostly a combination of what rust-mode does in Emacs and my attempts at respecting cosmetic issues like line length as per a hazy recollection of [the recommendations aturon compiled](https://aturon.github.io/). If you prefer, I'll run rustfmt over things before declaring this ready for a merge.